governance/delegation-interop.md

```diff
@@ -0,0 +1,65 @@
+
+However, pushing all of the logic onto the `GovToken` proves to be problematic given the potential flood of messages that may be out of sequence or have many smaller transfer events
+prior to a single delegation event. The following section aims to outline a solution to correct this as well as ensure that relayers still have the proper incentives to handle these
+messages.
+
+# Alternatives Considered
+
+When originally considering this problem, the idea was to create a hook during the `afterTokenTransfer` event. As alluded to above, this do not provide some of the desired functionality.It also makes a few assumptions about the behaviors of delegates that receive tokens and stifles their ability to create partial delegations in a low-cost manner. Moreover, if there was
+a want to later batch updates of voting power to mainnet it would be costly to those operating governance.
+
+Another active consideration is the use of block timestamp of each L2 instead of block number while implementing this new solution. The primary reason for block numbers to be used in
+its stead, stems from a concern of manipulation by sequencers and the possible delays that might need to be enforced between events to ensure correctness.
+
+# Proposed Solution
+
+## Rationale
+
+Due to the fact that the `GovernanceDelegation` contract now has awareness of the block numbers and chain IDs of all of the voting power checkpoints each L2 can now be responsible for
+their own checkpointing and thus make as many updates necessary without cost to relayers. From an implementation standpoint many of the changes are quite small and should minimize the
+risk of complication of an already intricate interface.
+
+The main goal is to ensure that the incentives are set up correctly so that relayers and indexers can manage and provide users the latest delegation changes when exercising their voting
+power.
+
+## Implementation
+
+Again, the implementation involves updates the Governor that will ensure that for voting power that has been migrated can be retrieved and updated based on the state on each of the L2 chains. It accomplishes this by keeping track of the token nonce and block numbers of each checkpoint in the mainnet contract. Instead of directly sending messages between the different instances, Governors on non-OP mainnet chains will now emit events regarding these blocks in the their checkpoint updates.
+
+The new storage values will be used to keep track of both the voting power that has been accumulated on the native chain as well as others thus not breaking backwards compatibility for
+non-migrated tokens. Indexers will now need to be aware of the events being emitted on both OP mainnet and the other L2's that implement the `GovToken` to show checkpoints that have yet
```

Comment:
So indexers need to watch every possible L2 that has governance token delegation? Just trying to understand the flow:
1. Event happens on L2 chain, changing delegation.
2. `GovToken` contract emits event in its checkpoint update.
3. `Governor` contract is aware of checkpoint updates(I'm not totally clear on how this works, is it using interop messaging to do this?)
4. `Governor` updates state and checkpoints and this info is indexed.
5. But indexer is _also_ listening for the L2 event emission because it should know about checkpoints from all L2s even if they are not on mainnet yet?

What happens if there is a delegation change on an L2 but it hasn't yet been submitted to mainnet? What is the source of truth if there is a vote in that scenario? It could be problematic if the onchain source of truth doesn't match the indexer/offchain source of truth, right? I think some sequence diagrams would be helpful here to visualize the various components and how they interact.

Reply:
I will include some diagrams in the document. But your understanding is correct, the source of truth is always the L2 but the block number inclusion allows votes to be counted on Mainnet for up to N specified blocks.